src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java

```diff
@@ -995,6 +994,18 @@ public boolean connect(SocketAddress remote) throws IOException {
         }
     }
 
+    @Override
+    public boolean connect(SocketAddress remote) throws IOException {
+        SocketAddress sa = checkRemote(remote);
+        if (!SocketConnectEvent.enabled()) {
+            return implConnect(sa);
+        }
+        long start = SocketConnectEvent.timestamp();
+        boolean connected = implConnect(sa);
+        SocketConnectEvent.offer(start, connected, sa);
+        return connected;
```

Comment:
It would be useful if the JBS or PR could say what the intent it for SocketChannels that are configured non-blocking. I assume that implConnect will execute in <20ms (the threshold in the JFR config) so no event will be ever be recorded when configured non-blocking. It would be possible to save the timestamp when connecting, and then use in finishConnect but that does depend on timely usage.

Reply:
The non-blocking case is poorly handled.  I'll update the JBS.  The untimely use of finishConnect would cause an artificially bad looking event duration which might be a bit misleading.

Reply:
> The non-blocking case is poorly handled. I'll update the JBS. The untimely use of finishConnect would cause an artificially bad looking event duration which might be a bit misleading.

My preference for now would be to focus on the blocking case. Why did my application pause for 30s? The other part to this will be the lookup methods in InetAddress as sometimes the apparent hang isn't establishing the connection but the DNS lookup.